include/seqan3/alphabet/union_alphabet.hpp

```diff
@@ -0,0 +1,186 @@
+constexpr auto at_least_type() {
+    if constexpr(size <= 1) {
+        return bool{};
+    } else if constexpr(size <= 255u){
+        return uint8_t{};
+    } else if constexpr(size <= 65535u){
+        return uint16_t{};
+    } else if constexpr(size <= 4294967295u){
+        return uint32_t{};
+    } else {
+        return uint64_t{};
+    }
+}
+} // namespace seqan::detail
+
+template <typename first_alphabet_type, typename ...alphabet_types>
+    requires alphabet_concept<first_alphabet_type> && (alphabet_concept<alphabet_types> && ...)
+struct union_alphabet
+{
+    /* types */
+    static constexpr size_t value_size = (alphabet_types::value_size + ... + first_alphabet_type::value_size);
+    using char_type = typename first_alphabet_type::char_type;
+    using integral_type = decltype(detail::at_least_type<value_size>());
+
+protected:
+    using tuple_t = std::tuple<first_alphabet_type, alphabet_types...>;
+    using index_sequence_t = decltype(std::make_index_sequence<std::tuple_size_v<tuple_t>>{});
+    static constexpr auto sizes = detail::alphabet_sizes<tuple_t>(index_sequence_t{});
+    static constexpr auto alphabets = detail::alphabet_objects<tuple_t>(index_sequence_t{});
+
+public:
+    integral_type value;
+
+    constexpr operator integral_type() const
+    {
+        return value;
+    }
+
+    constexpr char_type to_char() const
+    {
+        auto sizes_it = sizes.begin();
+        auto sizes_it_end = sizes.end();
+        auto alphabets_it = alphabets.begin();
+        auto size_start = 0;
+        auto size_end = 0;
+
+        std::optional<char_type> result;
+
+        while(sizes_it != sizes_it_end && !result.has_value()) {
+            size_end += *sizes_it;
+            std::visit([&](auto alphabet){
+                if(value < size_end) {
+                    result = alphabet.from_integral(value - size_start).to_char();
+                }
+            }, *alphabets_it);
+
+            size_start = size_end;
+            sizes_it++;
+            alphabets_it++;
+        }
+
+        return result.value_or(static_cast<char_type>(0));
+    }
+
+    constexpr integral_type to_integral() const
+    {
+        return value;
+    }
+
+    constexpr union_alphabet & from_char(char_type const c)
+    {
+        auto sizes_it = sizes.begin();
+        auto sizes_it_end = sizes.end();
+        auto alphabets_it = alphabets.begin();
+        auto size_start = 0;
+
+        bool found = false;
+
+        while(sizes_it != sizes_it_end && !found) {
+            std::visit([&](auto alphabet){
+                alphabet.from_char(c);
+                if(alphabet.to_char() == c) {
+                    value = size_start + alphabet.to_integral();
+                    found = true;
+                }
+            }, *alphabets_it);
+
+            size_start += *sizes_it;
+            sizes_it++;
+            alphabets_it++;
+        }
+
+        if(!found) {
+            value = 0;
+        }
+
+        return *this;
+    }
```

Comment:
respect for the metaprogramming, but couldn't this be done a little simpler? e.g. by recursing of the list of types and just check if `alphabet{}.from_char(c).to_char() == c`?

just a thought...